### DISCUSSION TITLE
Pass netrc to external configuration service

### DISCUSSION THREAD

**Discussion by pinpox** [woodpecker-ci/woodpecker#2268] [Request a Feature] [answered]
The [external configuration API](https://woodpecker-ci.org/docs/administration/external-configuration-api) currently does not get passed any credentials for repositories. 

This makes it impossible to use external configuration services with private repositories when the service needs access to the repository's contents.

As an example: I have a multi-user instance running, on which some users have private repositories. The instance uses my [configuration service](https://github.com/pinpox/woodpecker-flake-pipeliner/) for nix flakes, which needs to read a file contained in the repos itself. This is currently not possible for private repositories as neither the credentials nor the repo's contents itself are POST'ed to the external configuration service. 

My proposal: Extend the schema of the JSON that gets POST'ed to the configuration service with the contents of the netrc/credentials that are used to clone the repositories on the runners.

With this, the external configuration service migth also be a viable workaround for the [proposed pipeline compile step](https://github.com/woodpecker-ci/woodpecker/discussions/2254)

**Comment by anbraten** [comment, marked as answer]
I like the idea. From a security perspective it should also be fine as the instance admin sets the url of the config service and needs to make sure he trusts the config service

**Reply by pinpox** [reply]
I took a quick look at the code. I think it's just a matter of adding a Netrc to the requestStructure  [here](https://github.com/woodpecker-ci/woodpecker/blob/35baaa8224e2ecee4c13fdf2c23713d553124d74/server/plugins/config/http.go#L38-L42) so it looks like this

```go
type requestStructure struct {
	Repo          *model.Repo     `json:"repo"`
	Pipeline      *model.Pipeline `json:"pipeline"`
	Configuration []*config       `json:"configs"`
	Netrc         *model.Netrc    `json:"netrc"`
}
```

The struct is filled [here](https://github.com/woodpecker-ci/woodpecker/blob/35baaa8224e2ecee4c13fdf2c23713d553124d74/server/plugins/config/http.go#L63C24-L63C24), but I havn't been able to figure out yet how to get a Netrc in there. Can all the `Netrc()` method from there somehow?

**Reply by anbraten** [reply]
The forge has a function `func (c *config) Netrc(u *model.User, _ *model.Repo) (*model.Netrc, error) {` which you can use if you can provide a repo and user.

**Reply by pinpox** [reply]
But we have no `forge` in that function, should I just pass it in as additional argument? Or pass in the netrc directly?

**Reply by anbraten** [reply]
I would use sth like:

```go
netrc, err := cf.forge.Netrc(cf.user, cf.repo)
			if err != nil {
				return nil, fmt.Errorf("could not get Netrc data from forge: %w", err)
			}
			newConfigs, useOld, err := cf.configExtension.FetchConfig(fetchCtx, cf.repo, cf.pipeline, files, netrc)
			if err != nil {
				log.Error().Err(err).Msg("could not fetch config via http")
				return nil, fmt.Errorf("could not fetch config via http: %w", err)
			}
```

in `server/forge/configFetcher.go` and pass the netrc data from there.

**Reply by pinpox** [reply]
I opened a PR adding it https://github.com/woodpecker-ci/woodpecker/pull/2310
Not sure if the error handling is optimal, let me know what needs to change. 
I'm also in the matrix channel for faster communication if you prefer

**Reply by anbraten** [reply]
Thanks for the PR looks pretty simple and clean to me.
